File: summary/ruby/15_k_means.rb
def k_means(data, k)
  centroids = data.sample(k)
  clusters = {}

  loop do
    clusters = data.group_by do |point|
      centroids.min_by { |centroid| euclidean_distance(point, centroid) }
    end

    new_centroids = clusters.map do |_, points|
      points.transpose.map { |dimension| dimension.sum.to_f / dimension.size }
    end

    break if new_centroids == centroids

    centroids = new_centroids
  end

  return clusters
end

def euclidean_distance(point1, point2)
  sum_of_squares = point1.zip(point2).map { |a, b| (a - b) ** 2 }.sum
  Math.sqrt(sum_of_squares)
end

# Example usage:
data = [[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]]
k = 2
clusters = k_means(data, k)
clusters.each_with_index do |(centroid, points), index|
  puts "Cluster #{index + 1}: Centroid -> #{centroid}, Points -> #{points.inspect}"
end

File: summary/elixir/15_k_means.exs
defmodule KMeans do
  def cluster(data, k) do
    centroids = Enum.take_random(data, k)
    do_cluster(data, centroids)
  end

  defp do_cluster(data, centroids) do
    clusters = Enum.group_by(data, &closest_centroid(&1, centroids))
    new_centroids = Enum.map(clusters, fn {_, points} -> calculate_centroid(points) end)

    if new_centroids == centroids do
      clusters
    else
      do_cluster(data, new_centroids)
    end
  end

  defp closest_centroid(point, centroids) do
    Enum.min_by(centroids, &euclidean_distance(point, &1))
  end

  defp calculate_centroid(points) do
    points
    |> Enum.zip()
    |> Enum.map(&(Enum.sum(&1) / length(&1)))
  end

  defp euclidean_distance(point1, point2) do
    Enum.zip(point1, point2)
    |> Enum.map(fn {a, b} -> :math.pow(a - b, 2) end)
    |> Enum.sum()
    |> :math.sqrt()
  end
end

# 使用例
data = [[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]]
k = 2
clusters = KMeans.cluster(data, k)

Enum.each(clusters, fn {centroid, points} ->
  IO.puts("クラスター: 中心点 -> #{inspect(centroid)}, ポイント -> #{inspect(points)}")
end)
